k8s-tools/manage-cluster

```diff
@@ -558,16 +636,55 @@ function get_master_ips() {
   done
 }
 
+
 function open_shell() {
   # Open a shell using the KubeSpray image, which runs as root
-  local keyfile_path="$(get_key_file_path)"
+
+  local cluster_ks_version=$(_get_ks_stamp_version)
+  if [[ $# > 1 ]]; then
+    usage_error
+  elif [[ $# == 1 ]]; then
+    if [[ "${cluster_ks_version}" != "${1}" ]]; then
+      log "++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++"
+      log "WARNING:  Opening shell with KubeSpray version ${1} but template/cluster was created with version ${cluster_ks_version}"
+      log "++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++"
+      cluster_ks_version="${1}"
+    fi
+  fi
+  # If no version is set, fall back to default
+  if [[ -z "${cluster_ks_version}" ]]; then
+    cluster_ks_version="${DefaultKubesprayVersion}"
+  fi
+
   # We initialize the shell with the SSH credentials for the cluster.  This way,
   # the user can directly run ansible playbooks from it.  We also initialize
-  local init_cmds="eval \\\$(ssh-agent -s) ; ssh-add -k '${keyfile_path}';"
+  local keyfile_path="$(get_key_file_path)"
+  local init_cmds=(
+    "eval \\\$(ssh-agent -s); "
+    "ssh-add -k '${keyfile_path}';"
+    # if the checked-out version of Kubespray is different from the one requested, 
+    "[[ \\\$(cd \\\${KUBESPRAY_DIR} && git describe --tags) != ${cluster_ks_version} ]] && manage-cluster.py \\\${KUBESPRAY_DIR} --target-version ${cluster_ks_version} checkout --no-update-requirements")
+
+  ## Note that at the moment we're not installing the kubespray requirements
+  # when opening the shell to avoid excessively slowing down the start-up process.
+
+  docker_run_ks /bin/bash -c "/bin/bash --init-file <(echo \"${init_cmds[*]}\")" || true
+}
+
 
-  docker_run_ks /bin/bash -c "/bin/bash --init-file <(echo \"${init_cmds}\")" || true
+function list_supported_ks_versions() {
+  debug_log "==== list_supported_ks_versions ===="
+  docker_cmdline=(docker run
+                    -it --rm
+                    -e PYTHONPATH=/usr/local/bin/  
+                    --entrypoint /usr/bin/python3
+                    "${KsImage}"
+                    -c 'mc = __import__("manage-cluster"); print( "\n".join( ("{}\t{}".format(*row) for row in mc.VersionTable) ) )')
```

Comment:
I would like to add a table header, something like: **KubeSpray  -- Kubernetes**